src/zino/tasks/bgpstatemonitortask.py

```diff
@@ -0,0 +1,410 @@
+            else:
+                event = self.state.events.get(self.device.name, data.peer_remote_address, BGPEvent)
+                if event and event.operational_state != "established":
+                    self._bgp_external_reset(data)
+                    _logger.debug(f"BGP session up for {self.device_state.name}: {index}")
+        else:
+            saved_bgp_peer_admin_status = self.device_state.bgp_peer_admin_status.get(data.peer_remote_address, None)
+
+            if not saved_bgp_peer_admin_status:
+                self.device_state.bgp_peer_admin_status[data.peer_remote_address] = "unknown"
+                saved_bgp_peer_admin_status = "unknown"
+
+            if data.peer_admin_status in ["stop", "halted"]:
+                if saved_bgp_peer_admin_status != data.peer_admin_status:
+                    self._bgp_admin_down(data)
+                    _logger.debug(
+                        f"Router {self.device_state.name} peer {data.peer_remote_address} AS {data.peer_remote_as} "
+                        f"admin-down"
+                    )
+            # peer_admin_status is start or running
+            else:
+                if saved_bgp_peer_admin_status != data.peer_admin_status:
+                    self._bgp_admin_up(data)
+
+                saved_bgp_peer_oper_state = self.device_state.bgp_peer_oper_states.get(data.peer_remote_address, None)
+
+                if not saved_bgp_peer_oper_state:
+                    self.device_state.bgp_peer_oper_states[data.peer_remote_address] = "established"
+                    saved_bgp_peer_oper_state = "established"
+
+                if saved_bgp_peer_oper_state == "established":
+                    # First verify that we've been up longer than the required time before we flag it as an alert
+                    if uptime > TIME_BEFORE_OPER_DOWN_ALERT:
+                        self._bgp_oper_down(data)
+                        _logger.debug(
+                            f"Router {self.device_state.name} peer {data.peer_remote_address} AS {data.peer_remote_as} "
+                            f"is {data.peer_state} (down)",
+                        )
+                    else:
+                        _logger.debug(
+                            f"Router {self.device_state.name} peer {data.peer_remote_address} AS {data.peer_remote_as} "
+                            f"is {data.peer_state} (down), but uptime = {uptime}",
+                        )
+
+        # Update all states
+        self.device_state.bgp_peer_oper_states[data.peer_remote_address] = data.peer_state
+        self.device_state.bgp_peer_admin_status[data.peer_remote_address] = data.peer_admin_status
+        self.device_state.bgp_peer_up_times[data.peer_remote_address] = data.peer_fsm_established_time
+
+        # Add new peers if found
+        if data.peer_remote_address not in self.device_state.bgp_peers:
+            self.device_state.bgp_peers.append(data.peer_remote_address)
```

Comment:
You asked for suggestions for a refactor.  Without diving very deep into the logic here, it could at least be easily broken down into multiple methods using a simple refactor operation (usually even available as automated tools from your IDE).

One very quick suggestion is first to extract this into a module-level constant:
```python
BUGGY_REMOTE_ADDRESSES = [
    # Bug in JunOS -- info from IPv6 BGP sessions spill over
    "0.0.0.0",
    # Bug in earlier Cisco IOS, info from elsewhere (IPv6?) spills over
    "32.1.7.0",
]
```

Then just factor out some of the branches like so:

```suggestion
    def _update_single_bgp_entry(self, oid: OID, row: dict[str, Any], local_as: int, uptime: int):
        data = BaseBgpRow(**row)
        if data.peer_remote_address in BUGGY_REMOTE_ADDRESSES:
            return

        # Internal bgp sessions are not observed
        if local_as == data.peer_remote_as:
            return

        if data.peer_state == "established":
            self._update_established_peer(data, uptime, oid)
        else:
            self._update_nonestablished_peer(data, uptime)

        # Update all states
        self.device_state.bgp_peer_oper_states[data.peer_remote_address] = data.peer_state
        self.device_state.bgp_peer_admin_status[data.peer_remote_address] = data.peer_admin_status
        self.device_state.bgp_peer_up_times[data.peer_remote_address] = data.peer_fsm_established_time

        # Add new peers if found
        if data.peer_remote_address not in self.device_state.bgp_peers:
            self.device_state.bgp_peers.append(data.peer_remote_address)

    def _update_established_peer(self, data, uptime, oid):
        index = f"{self.device.name},{oid}"

        saved_bgp_peer_up_time = self.device_state.bgp_peer_up_times.get(data.peer_remote_address, None)
        if saved_bgp_peer_up_time and uptime >= saved_bgp_peer_up_time > data.peer_fsm_established_time:
            self._bgp_external_reset(data)
            _logger.debug(f"Noted external reset for {self.device_state.name}: {index}")
        else:
            event = self.state.events.get(self.device.name, data.peer_remote_address, BGPEvent)
            if event and event.operational_state != "established":
                self._bgp_external_reset(data)
                _logger.debug(f"BGP session up for {self.device_state.name}: {index}")

    def _update_nonestablished_peer(self, data, uptime):
        saved_bgp_peer_admin_status = self.device_state.bgp_peer_admin_status.get(data.peer_remote_address, None)
        if not saved_bgp_peer_admin_status:
            self.device_state.bgp_peer_admin_status[data.peer_remote_address] = "unknown"
            saved_bgp_peer_admin_status = "unknown"
        if data.peer_admin_status in ["stop", "halted"]:
            if saved_bgp_peer_admin_status != data.peer_admin_status:
                self._bgp_admin_down(data)
                _logger.debug(
                    f"Router {self.device_state.name} peer {data.peer_remote_address} AS {data.peer_remote_as} "
                    f"admin-down"
                )
        else:
            self._update_peer_with_admin_status_start_or_running(data, saved_bgp_peer_admin_status, uptime)

    def _update_peer_with_admin_status_start_or_running(self, data, saved_bgp_peer_admin_status, uptime):
        if saved_bgp_peer_admin_status != data.peer_admin_status:
            self._bgp_admin_up(data)
        saved_bgp_peer_oper_state = self.device_state.bgp_peer_oper_states.get(data.peer_remote_address, None)
        if not saved_bgp_peer_oper_state:
            self.device_state.bgp_peer_oper_states[data.peer_remote_address] = "established"
            saved_bgp_peer_oper_state = "established"
        if saved_bgp_peer_oper_state == "established":
            # First verify that we've been up longer than the required time before we flag it as an alert
            if uptime > TIME_BEFORE_OPER_DOWN_ALERT:
                self._bgp_oper_down(data)
                _logger.debug(
                    f"Router {self.device_state.name} peer {data.peer_remote_address} AS {data.peer_remote_as} "
                    f"is {data.peer_state} (down)",
                )
            else:
                _logger.debug(
                    f"Router {self.device_state.name} peer {data.peer_remote_address} AS {data.peer_remote_as} "
                    f"is {data.peer_state} (down), but uptime = {uptime}",
                )
```

The `oid` argument to `_update_single_bgp_entry` appears to be there exclusively for constructing some debug log statements, which looks a bit weird.  Maybe it could be obtained some other way, but that would possibly require a deeper analysis than I've committed to :)

Reply:
Very good catch, with the oid, I realized it was a mistake on my part and I used the wrong value for logging